### ISSUE TITLE
Rule Request: weak, unowned in noescape closures

### ISSUE THREAD

**Issue by Igor-Palaguta** [realm/SwiftLint#2393]
### Rule Request

noescape_captures

1. Why should this rule be added? Share links to existing discussion about what
the community thinks about this.

No need to write weak, unowned in capture list in noescape closure.

2. Provide several examples of what _would_ and _wouldn't_ trigger violations.
Would trigger:

```swift
let modifiedArray = array.map { [weak self] value in
    self?.modify(value) ?? ""
}

let modifiedArray2 = array.map { [unowned self] value in
    self.modify(value)
}
```

Would not trigger:

```swift
let modifiedArray = array.map { value in
    modify(value)
}
```

3. Should the rule be configurable, if so what parameters should be configurable?

- Probably we can always require not to write self inside noespace closures. As there is no need for it

- Probably we should allow/disallow to use capture list without ownership

```swift
let modifiedArray = array.map { [modifier = transformer.operations.modifier] value in
    modifier.modify(value)
}
```

4. Should the rule be opt-in or enabled by default? Why?
   
For weak and unowned - Yes

**Comment by mortenholmgaard** [comment]
@jpsim Please have a look at this request. I know it is complex but this would help ensure a lot less memory leaks and crashes in swiftlint apps!

**Comment by jpsim** [comment]
The big challenge with a rule like this is that we can't rely on the pre-typechecked AST to lint this reliably, because so many `map` functions operate on asynchronous streams of data (e.g. RxSwift, Combine), and in those cases `map` takes escaping closures.

So we could either:

1) write a rule with a lot of false positives in codebases with many uses of `map` that have escaping closures
2) write an analyzer rule that operates on the post-typechecked AST that could reliably detect when `map` has an escaping or non-escaping closure, but this would be just as slow and cumbersome to lint as all our other analyzer rules, but it would be correct

**Comment by mortenholmgaard** [comment]
To me it sounds like option 2 is the only viable solution even thought it will be slow.
